src/main/java/emissary/parser/NIOSessionParser.java

```diff
@@ -68,49 +84,76 @@ public void setChunkSize(int value) {
      *
      * @param data the byte array to (re)load or null if one should be created
      * @return the byte array of data
+     * @throws ParserException in cases where a new array can't be read.
      */
-    protected byte[] loadNextRegion(byte[] data) {
-        logger.debug("loadOrFillNextRegion: data.length = {}, chunkSize = {}, chunkStart = {}",
-                data == null ? -1 : data.length, chunkSize, chunkStart);
+    protected byte[] loadNextRegion(byte[] data) throws ParserException {
+        logger.debug("loadNextRegion(): data.length = {}, maxChunkSize = {}, chunkStart = {}, writeOffset = {}",
+                data == null ? -1 : data.length, maxChunkSize, chunkStart, writeOffset);
 
-        // Position before checking remaining
-        try {
-            channel.position(chunkStart);
-        } catch (IOException iox) {
-            logger.error("Unable to seek to {}", chunkStart, iox);
-            return null;
+        if (!channel.isOpen()) {
+            logger.debug("loadNextRegion(): channel closed");
+            throw new ParserEOFException("Channel is closed, likely completely consumed");
+        }
+
+        // Position ourselves in the input stream at the beginning of the next session when starting a new chunk.
+        if (writeOffset == 0) {
+            // Position before checking remaining
+            try {
+                channel.position(chunkStart);
+            } catch (IOException ioe) {
+                throw new ParserException("Exception positioning channel to " + chunkStart, ioe);
+            }
         }
 
         // Optionally create the array or recreate if old is too small
-        if (data == null || data.length < chunkSize) {
-            data = new byte[chunkSize];
+        if (data == null) {
+            logger.debug("allocating new byte[] of size {}", minChunkSize);
+            data = new byte[minChunkSize];
+        }
+
+        if (writeOffset >= data.length) {
+            // grow the byte buffer to accomodate more data
+            int newSize = data.length + chunkAllocationIncrement;
+            if (newSize > maxChunkSize) {
```

Comment:
So the chuckAllocationAgreement (100MB) is already greater than the default maxChuckSize (40MB). 

This should always evaluate to true in the default configuration. Is that intentional?